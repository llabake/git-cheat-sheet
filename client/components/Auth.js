/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { Component } from 'react';
import { connect } from "react-redux";
import toastr from 'toastr';

import TextInput from "./Common/TextInput";
import { userInputValidator } from "../helpers/utils";
import { authUser } from "../actions/authAction";

export class Auth extends Component {
  state = {
    user: {
      username: "",
      password: "",
    },
    errors: {},
    isValid: false,
    saving: false,
    showSignUp: true,
  };

  handleChange = (event) => {
    const field = event.target.name;
    const { user, errors } = this.state;
    user[field] = event.target.value;
    errors[field] = '';
    this.setState({ user, errors });
  };

  validate = () => {
    const { errors, isValid } = userInputValidator(this.state);
    this.setState({ isValid, errors });
    return isValid;
  };

  saveUser = (event) => {
    event.preventDefault();
    if (!this.validate()) {
      return;
    }
    this.setState({ saving: true });
    this.props.authUser(this.state.showSignUp ? 'signup' : 'login', this.state.user)
      .then(() => {})
      .catch((error) => {
        toastr.error(error);
        this.setState({ saving: false });
      });
  };

  toggleShowSignin = () => {
    this.setState({ showSignUp: false });
  };

  toggleShowSignup = () => {
    this.setState({ showSignUp: true });
  };

  render() {
    const { errors, user: { username, password }, showSignUp } = this.state;
    return (
      <div>
        <ul>
          <li onClick={this.toggleShowSignup}>
            <i data-tip data-for="register" className="material-icons">
              person_add
            </i>
          </li>
          <li onClick={this.toggleShowSignin}>
            <i data-tip data-for="login" className="material-icons">
              account_circle
            </i>
          </li>
        </ul>

        <form onSubmit={this.saveUser}>
          <TextInput
            error={errors.username}
            type="text"
            onChange={this.handleChange}
            value={username}
            icon="account_circle"
            name="username"
            placeholder="Username"
          />
          <TextInput
            error={errors.password}
            type="password"
            onChange={this.handleChange}
            value={password}
            icon="lock"
            name="password"
            placeholder="Password"
          />
          <button
            type="submit"
            className="register-btn"
            value="register">
            { showSignUp ? "Register" : "Login" }
          </button>
        </form>
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  error: state.error,
  user: state.userReducer,
});

export const mapDispatchToProps = dispatch => ({
  authUser: (type, user) => dispatch(authUser(type, user)),
});
export default connect(mapStateToProps, mapDispatchToProps)(Auth);
