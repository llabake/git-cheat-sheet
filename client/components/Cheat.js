import React, { Component } from 'react';
import { CopyToClipboard } from 'react-copy-to-clipboard';

export default class Cheat extends Component {
  state = {
    copied: false,
  };

  clicked = () => {
    this.setState({ copied: true }, () => {
      setTimeout(() => this.setState({ copied: false }), 1000);
    });
  };

  render() {
    const { cheat: { description, commands } } = this.props;
    const { copied } = this.state;
    const [keyword, ...body] = commands.split(' ');
    const remainder = body.join(' ');
    return (
      <div className="header">
        <span>{description}</span>
        <span style={{ display: `${copied ? 'inline' : 'none'}` }} className="cheat-copied">
          Copied!
        </span>
        <pre>
          $&nbsp;
          <code>
            <CopyToClipboard text={commands} onCopy={this.clicked}>
              <span>
                <span className="first">{keyword}</span>
                <span>
                &nbsp;
                  {remainder}
                </span>
              </span>
            </CopyToClipboard>
          </code>
        </pre>
      </div>
    );
  }
}
