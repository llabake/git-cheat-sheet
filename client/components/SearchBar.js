import React from 'react';
import { connect } from 'react-redux';

import { searchCategory } from "../actions/categoryAction";

export const SearchBar = ({ searchCategory: search }) => (
  <div className="row">
    <div className="input-field col offset-s3 s6">
      <i style={{ top: '1rem' }} className="material-icons prefix">search</i>
      <input
        onChange={ event => search(event.target.value) }
        id="search"
        type="text"
      />
      <label className="search-placeholder" htmlFor="search" style={{ textAlign: 'center' }}>
        The Awesome Git Cheat Sheet
      </label>
    </div>
  </div>
);

export default connect(null, { searchCategory })(SearchBar);
