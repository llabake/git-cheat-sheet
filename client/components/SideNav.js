import React, { Component } from 'react';
import Logo from '../public/images/giticonw.png';

import Auth from './Auth';

export default class SideNav extends Component {
  componentDidMount() {
    this.initSideNav();
  }

  componentDidUpdate() {
    this.initSideNav();
  }

  initSideNav = () => {
    document.addEventListener('DOMContentLoaded', () => {
      M.Sidenav.init(document.querySelectorAll('.sidenav'), null);
    });
  };

  render() {
    return (
      <div className="col">
        <a href="#!" data-target="slide-out" className="sidenav-trigger">
          <i style={{ marginTop: '2rem' }} className="material-icons">menu</i>
        </a>
        <div id="slide-out" className="sidenav">
          <Auth />
        </div>
      </div>
    );
  }
}
