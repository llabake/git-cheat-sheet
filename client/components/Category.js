import React from 'react';

import Cheat from './Cheat';

const Category = ({ category }) => (
  <div className="category col s4">
    <span className="category-title">{category.title}</span>
    <div className="category-cheat">
      {
        category.cheats.map(cheat => <Cheat key={cheat._id} cheat={cheat} />)
      }
    </div>
  </div>
);

export default Category;