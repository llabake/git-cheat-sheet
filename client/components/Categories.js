import React from 'react';
import { connect } from 'react-redux';

import Category from './Category';

export default connect(state => ({
  categories: state.categories,
  searchResults: state.search,
}))(({ categories, searchResults }) => {
  const data = searchResults || categories;
  return (
    data.map(category => (
      <Category key={category._id} category={category} />
    ))
  );
});
