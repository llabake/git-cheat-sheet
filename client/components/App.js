import React, { Component } from 'react';
import SideNav from "./SideNav";
import Categories from "./Categories";
import SearchBar from "./SearchBar";

export default class App extends Component {
  render() {
    return (
      <div className="row">
        <SideNav />
        <div className="container">
          <SearchBar />
          <Categories />
        </div>
      </div>
    )
  }
}
