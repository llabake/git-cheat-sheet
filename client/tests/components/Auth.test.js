import React from 'react';
import { shallow } from 'enzyme';

import { Auth, mapDispatchToProps } from '../../components/Auth';

const setup = () => {
  const props = {
    user: {
      username: "",
      password: "",
    },
    errors: {},
    isValid: false,
    saving: false,
    showSignUp: true,
    handleChange: jest.fn(() => new Promise(resolve => resolve())),
    validate: jest.fn(() => new Promise(resolve => resolve())),
    saveUser: jest.fn(() => new Promise(resolve => resolve())),
    signOut: jest.fn(() => new Promise(resolve => resolve())),
    toggleShowSignin: jest.fn(() => new Promise(resolve => resolve())),
    toggleShowSignup: jest.fn(() => new Promise(resolve => resolve())),
  };

  const wrapper = shallow(<Auth {...props} />);

  return { wrapper, props };
};

describe('<Auth />', () => {
  const { wrapper } = setup();

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
	
  it('handle change', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleChange');
    const event = {
      preventDefault: jest.fn(),
      target: {
        username: 'labake',
      },
    };
    wrapper.instance().handleChange(event);
    expect(spy).toHaveBeenCalled();
  });

  it('handles validation', () => {
    const spy = jest.spyOn(wrapper.instance(), 'validate');
    const event = {
      preventDefault: jest.fn(),
      target: {
        username: 'la',
        password: 'p',
      },
    };
    wrapper.setState({
      errors: {
        username: 'too short',
        password: 'too short',
      },
    });
    wrapper.instance().validate(event);
    expect(spy).toHaveBeenCalled();
  });
	
  it('saves a user', () => {
    const spy = jest.spyOn(wrapper.instance(), 'saveUser');
    const event = {
      preventDefault: jest.fn(),
      target: {
        username: 'labake',
        password: 'password',
      },
		};
		wrapper.setState({
      saving: true,
    });
    wrapper.instance().saveUser(event);
    expect(spy).toHaveBeenCalled();
  });
	
  it('toggles sign in', () => {
    const spy = jest.spyOn(wrapper.instance(), 'toggleShowSignin');
    wrapper.setState({
      showSignUp: false,
    });
    wrapper.instance().toggleShowSignin();
    expect(spy).toHaveBeenCalled();
	});
	
	it('toggles sign up', () => {
    const spy = jest.spyOn(wrapper.instance(), 'toggleShowSignin');
    wrapper.setState({
      showSignUp: true,
    });
    wrapper.instance().toggleShowSignup();
    expect(spy).toHaveBeenCalled();
  });
  
  it('sign out a user', () => {
    const spy = jest.spyOn(wrapper.instance(), 'signOut');
    const event = {
      preventDefault: jest.fn(),
		};
    wrapper.instance().signOut(event);
    expect(spy).toHaveBeenCalled();
	});
	
	it('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('authUser');
    expect(mapDispatchToProps(dispatch)).toHaveProperty('signOut');
    const { authUser, signOut } = mapDispatchToProps(dispatch);
    authUser('login', { username: 'james', password: 'pasdss'} );
    expect(dispatch).toHaveBeenCalled();
    signOut()
    expect(dispatch).toHaveBeenCalled();
  });
});
