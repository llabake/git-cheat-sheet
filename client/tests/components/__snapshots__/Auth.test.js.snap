// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Auth /> should render correctly 1`] = `
<div>
  <div
    style={
      Object {
        "margin": "2rem",
      }
    }
  >
    <strong>
      Enjoy the Cheat Sheet
    </strong>
    <i
      className="material-icons"
      data-for="logout"
      data-tip={true}
      onClick={[Function]}
    >
      lock
    </i>
  </div>
  <ReactTooltip
    className="tooltip"
    effect="solid"
    id="signup"
    insecure={true}
    resizeHide={true}
    wrapper="div"
  >
    <span>
      Signup
    </span>
  </ReactTooltip>
  <ReactTooltip
    className="tooltip"
    effect="solid"
    id="login"
    insecure={true}
    resizeHide={true}
    wrapper="div"
  >
    <span>
      Login
    </span>
  </ReactTooltip>
  <ReactTooltip
    className="tooltip"
    effect="solid"
    id="logout"
    insecure={true}
    resizeHide={true}
    wrapper="div"
  >
    <span>
      Logout
    </span>
  </ReactTooltip>
</div>
`;
